bson/bsoncodec/map_codec.go

```diff
@@ -204,3 +213,80 @@ func clearMap(m reflect.Value) {
 		m.SetMapIndex(k, none)
 	}
 }
+
+func (mc *MapCodec) encodeKey(val reflect.Value) (string, error) {
+	if mc.EncodeKeysWithStringer {
+		return fmt.Sprint(val), nil
+	}
+
+	// keys of any string type are used directly
+	if val.Kind() == reflect.String {
+		return val.String(), nil
+	}
+	// KeyMarshalers are marshaled
+	if tm, ok := val.Interface().(KeyMarshaler); ok {
```

Comment:
nit: I assume `tm` was for `TextMarshaler`. Change to `km`.

Reply:
Should we be checking if a pointer to value implements `KeyMarshaler` as well? Our standard registry lookup checks that for other interfaces we support like `Marshaler`.

Reply:
from what i can tell, map keys aren't addressable, so I didn't any of the pointer to value things.

Reply:
Yep, seems like `encoding/json` doesn't respect &T implementing `TextMarshaler` either. As discussed offline, this is likely because map keys and values can move around if the map is re-sized so any addresses to them would be invalidated.